zipline/finance/asset_restrictions.py

```diff
@@ -0,0 +1,215 @@
+import abc
+from numpy import vectorize
+from functools import partial, reduce
+import operator
+import pandas as pd
+from six import with_metaclass
+from collections import namedtuple
+from itertools import groupby
+
+from zipline.utils.enum import enum
+from zipline.utils.numpy_utils import vectorized_is_element
+from zipline.assets import Asset
+
+
+Restriction = namedtuple(
+    'Restriction', ['asset', 'effective_date', 'state']
+)
+
+
+RESTRICTION_STATES = enum(
+    'ALLOWED',
+    'FROZEN',
+)
+
+
+class Restrictions(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract restricted list interface
```

Comment:
Often for the ABC in a hierarchy, you want a sentence or two describing what an instance of that type represents abstractly.  Here, it would be something like: "ABC representing a set of stocks that an algorithm is restricted from trading.".  For bonus points, you could add a [`Methods`](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#class-docstring) section listing the names and signatures of the methods required for implementing the interface.